proxy/proxy.go

```diff
@@ -287,7 +309,141 @@ func (p *Proxy) createListener(local, upstream string) (*listener.Listener, erro
 		_ = s.Disconnect(ctx)
 	}
 
-	return listener.New(logWith, sdWith, p.config.Network, local, p.config.Unlink, connectionHandler, shutdownHandler)
+	listener, err := listener.New(logWith, sdWith, p.config.Network, local, p.config.Unlink, connectionHandler, shutdownHandler)
+	if err != nil {
+		return nil, err
+	}
+	return &ListenerServerPair{
+		Listener: listener,
+		Server:   s,
+	}, nil
+}
+
+// Go through all listeners and health check their servers
+// This should ideally by its own class with a healthcheck strategy
+// Due to timeline pressures we'll just use this basic scheme
+// to avoid repeated loops of Timeout on the redis client
+func (p *Proxy) checkConnections() {
+	duration := time.Duration(p.config.ServerHealthCheckSec) * time.Second
+	p.log.Debug("Inside checkConnections", zap.String("duration", duration.String()))
+	for {
+		time.Sleep(duration)
+		p.log.Debug("Just woke up to check connections")
+		keys := p.getListenerKeys()
+		c := make(chan string)
+		for _, key := range keys {
+			go p.checkSingleConnection(key, c)
+		}
+		for range keys {
+			<-c
+		}
+	}
+}
+
+// Health check a single host:port
+// Process:
+// - check out a connection from the server
+// - send a ping command
+// - if error, repeat for x times
+// - if kept on erroring:
+//   - replace listener if the upstreamConfigHost
+//   - remove listener if otherwise
+//
+// If the server repeatedly fails remove them unless they are the main upstream host
+// In that case, just try to recreate the connections
+// The listeners that get removed will be re-created only during the intercepts of
+// CLUSTER NODES commands sent by the client
+func (p *Proxy) checkSingleConnection(key string, c chan string) {
+	p.log.Debug("Inside checkSingleConnection", zap.String("server", key))
+	ls := p.getListenerServerPair(key)
+	p.log.Debug("getListenerServerPair returned", zap.String("ls!=nil", strconv.FormatBool(ls != nil)))
+	if ls != nil {
+		failing := false
+		messenger := messenger.WireMessenger{}
+		for i := 0; i < int(p.config.ServerHealthCheckThreshold); i++ {
+			time.Sleep(1 * time.Second)
+			failing = !p.pingServer(ls.Server, messenger)
+			p.log.Debug("Finished pinging server", zap.String("server", key))
+			if !failing {
+				break
+			}
```

Comment:
minor: to minimize cognitive dissonance, can we refactor to something like:

```
health := true
...
for i := 0; i < int(p.config.ServerHealthCheckThreshold); i++ {
			time.Sleep(1 * time.Second)
			health = p.pingServer(ls.Server, messenger)
			p.log.Debug("Finished pinging server", zap.String("server", key))
			if health {
				break
			}
}
if !health {
    ...
}
```

Reply:
oh yeah I see what you mean :)